File: coding-labs/senior-technical-screen/python/src/idempotent_command_handler.py
from collections.abc import Callable
from dataclasses import dataclass
from enum import Enum
from threading import Lock


class Status(Enum):
    CREATED = "created"
    REPLAYED = "replayed"


@dataclass(frozen=True)
class Command:
    request_id: str
    payload: str


@dataclass(frozen=True)
class Result:
    resource_id: str
    status: Status


class IdempotentCommandHandler:
    """Serializes local duplicates; production needs durable uniqueness."""

    def __init__(self, repository: Callable[[str], str]):
        if repository is None:
            raise ValueError("repository is required")
        self._repository = repository
        self._completed: dict[str, tuple[str, str]] = {}
        self._lock = Lock()

    def handle(self, command: Command) -> Result:
        if command is None:
            raise ValueError("command is required")
        request_id = self._normalize(command.request_id, "request_id")
        payload = self._normalize(command.payload, "payload")

        with self._lock:
            existing = self._completed.get(request_id)
            if existing is not None:
                stored_payload, resource_id = existing
                if stored_payload != payload:
                    raise RuntimeError("request ID already used with different content")
                return Result(resource_id, Status.REPLAYED)

            resource_id = self._repository(payload)
            if resource_id is None or not resource_id.strip():
                raise RuntimeError("repository returned an invalid result")
            self._completed[request_id] = (payload, resource_id)
            return Result(resource_id, Status.CREATED)

    @staticmethod
    def _normalize(value: str, field: str) -> str:
        if value is None or not value.strip():
            raise ValueError(f"{field} is required")
        return value.strip()

File: coding-labs/senior-technical-screen/java/src/main/java/IdempotentCommandHandler.java
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public final class IdempotentCommandHandler {
  public enum Status { CREATED, REPLAYED }

  public record Command(String requestId, String payload) {}

  public record Result(String resourceId, Status status) {}

  @FunctionalInterface
  public interface Repository {
    String create(String payload);
  }

  private record Stored(String payload, String resourceId) {}

  private final Repository repository;
  private final Map<String, Stored> completed = new HashMap<>();

  public IdempotentCommandHandler(Repository repository) {
    this.repository = Objects.requireNonNull(repository, "repository is required");
  }

  /** Serializes local duplicates; production requires a durable uniqueness boundary. */
  public synchronized Result handle(Command command) {
    if (command == null) throw new IllegalArgumentException("command is required");
    String requestId = normalize(command.requestId(), "requestId");
    String payload = normalize(command.payload(), "payload");

    Stored existing = completed.get(requestId);
    if (existing != null) {
      if (!existing.payload().equals(payload)) {
        throw new IllegalStateException("request ID already used with different content");
      }
      return new Result(existing.resourceId(), Status.REPLAYED);
    }

    String resourceId = repository.create(payload);
    if (resourceId == null || resourceId.isBlank()) {
      throw new IllegalStateException("repository returned an invalid result");
    }
    completed.put(requestId, new Stored(payload, resourceId));
    return new Result(resourceId, Status.CREATED);
  }

  private static String normalize(String value, String field) {
    if (value == null || value.isBlank()) throw new IllegalArgumentException(field + " is required");
    return value.trim();
  }
}
